Add feature flags documentation (#3968)
PostHog/posthog.com: contents/docs/integrate/server/ruby.mdx

})
```

### Feature flags

> PostHog's [Feature Flags](/docs/user-guides/feature-flags) allow you to safely deploy and roll back new features.

Reviewer: ```suggestion
PostHog's [feature flags](/docs/user-guides/feature-flags) enable you to safely deploy and roll back new features.
```

When using them with one of libraries, you should check if a feature flag is enabled and use the result to toggle functionality on and off in you application.

**How to check if a flag is enabled**

> **Note:** Whenever we face an error computing the flag, the library returns `Nil`, instead of `true` or `false`.

```c
posthog.is_feature_enabled('beta-feature', 'distinct id')
# returns True or False or Nil
```

**Example use case**

Here's how you might send different users a different version of your homepage, for example:

```c
def homepage(request):
    template = "new.html" if posthog.is_feature_enabled('new_ui', 'distinct id') else "old.html"
    return render_template(template, request=request)
```

> **Note:** Feature flags are persistent for users across sessions. Read more about feature flag persistence on our [dedicated page](/docs/user-guides/feature-flags#feature-flag-persistence).

**Get a flag value**

If you're using multivariate feature flags, you can also get not just whether the flag is enabled, but what value its enabled to.

> > **Note:** Whenever we face an error computing the flag, the library returns `Nil`, instead of `true` or `false` or a string variant value.

Reviewer: ```suggestion
> **Note:** Whenever we face an error computing the flag, the library returns `Nil`, instead of `true`, `false`, or a string variant value.
```

@neilkakkar Other places in the docs we say we return `None`, but here we say we return `Nil` instead. Does this make a difference, or is one incorrect?

Author: In `ruby`, the null value is `Nil`. In `python`, it's `None`, and in javascript, it's `undefined`, hence the 3 in the docs for the 3 different libraries :)

Reviewer: I love learning about code! Thanks :)

```c
posthog.get_feature_flag('beta-feature', 'distinct id')

> # returns string or True or False or Nil

Reviewer: As above RE `Nil` or `None`.

```

**Overriding server properties**

Sometimes, you might want to evaluate feature flags using properties that haven't been ingested yet, or were set incorrectly earlier. You can do so by setting properties the flag depends on with these calls.

For example, if the `beta-feature` depends on the `is_authorized` property, and you know the value of the property, you can tell PostHog to use this property, like so:

```c
posthog.get_feature_flag('beta-feature', 'distinct id', person_properties: {'is_authorized': True})
```

You can pass symbols or strings, the library treats them the same.

> The same holds for groups. If you have a group name `organisation`, you can add properties like so:

Reviewer: ```suggestion
The same holds for [groups](/manual/group-analytics). If you have a group name `organisation`, you can add properties like so:
```

```python
posthog.get_all_flags('distinct id', groups: {}, person_properties: {'is_authorized': True}, group_properties: {})
# returns dict of flag key and value pairs.
```

#### Local Evaluation

> **Note:** To enable local evaluation of feature flags you must also set a `personal_api_key` when configuring the integration, as described in the [Installation](#installation) section.

> **Note:** This feature requires version 2.0 of the library, which in turn requires a minimum PostHog version of 1.38

All feature flag evaluation requires an API request to your PostHog servers to get a response. However, where latency matters, you can evaluate flags locally. This is much faster, and requires two things to work:

1. The library must be initialised with a personal API key
2. You must know all person or group properties the flag depends on.

Then, the flag can be evaluated locally. The method signature looks exactly like above

```c
posthog.get_feature_flag('beta-feature', 'distinct id', person_properties: {'is_authorized': True})

> # returns string or Nil

Reviewer: Nil vs None

// Do something with feature flags here
```

### Group analytics

PostHog 1.31.0 introduced support for [group analytics](/docs/user-guides/group-analytics), which allows you to associate users and events with larger groups (teams, organizations, etc.). This feature requires a posthog-ruby version of `2.0.0` or above.

> > **Note: ** This is a paid feature and is not available on the open-source or free cloud plan. Learn more [here](/pricing).

Reviewer: ```suggestion
> **Note: ** This is a paid feature and is not available on the open-source or free cloud plan. [Learn more about PostHog pricing](/pricing).
```
